dhall/src/Dhall/Import.hs

```diff
@@ -1119,3 +1121,50 @@ instance Show ImportResolutionDisabled where
 assertNoImports :: MonadIO io => Expr Src Import -> io (Expr Src Void)
 assertNoImports expression =
     Dhall.Core.throws (traverse (\_ -> Left ImportResolutionDisabled) expression)
+
+{-| This function is used by the @--transitive@ flag of the
+    @dhall {freeze,format,lint}@ subcommands to determine which dependencies
+    to descend into
+-}
+dependencyToFile :: Status -> Import -> IO (Maybe FilePath)
+dependencyToFile status import_ = flip State.evalStateT status $ do
+    parent :| _ <- zoom stack State.get
+
+    child <- fmap chainedImport (hoist liftIO (chainImport parent import_))
+
+    let ignore = return Nothing
+
+    -- Only descend into "normal" imports
+    case importMode child of
+        RawText ->
+            ignore
+
+        Location ->
+            ignore
+
+        Code ->
+            case importType (importHashed child) of
+                Local filePrefix file -> do
+                    let descend = liftIO $ do
+                            path <- localToPath filePrefix file
+
+                            return (Just path)
+
+                    -- Only follow relative imports when modifying dependencies.
+                    -- Carefully note that we check the file prefix of the
+                    -- original import (before chaining), since the chained
+                    -- import will inherit the file prefix of the parent import.
+                    case importType (importHashed import_) of
+                        Local Here   _ -> descend
+                        Local Parent _ -> descend
+                        _              -> ignore
+
+                -- Don't transitively modify any other type of import
+                Remote{} ->
+                    ignore
+
+                Missing ->
+                    ignore
+
+                Env{} ->
+                    ignore
```

Comment:
[style|curiosity] why don't you use `_` to catch all these cases?

Reply:
It's mainly so that I get a warning if we ever extend the language to handle other import types